### PR TITLE
feat: add support for universe domain (phase 1)

### PR DESCRIPTION
**Pull request by bshaffer**
Requirements:

1. For all credentials: Create the `universe_domain` read property. The default value for the property is `googleapis.com`.
2. For Service Account Credentials: Read the `universe_domain` from the credentials file and expose the value through the property created above. If the credentials file does not have the universe_domain value, the property should use the default value `googleapis.com`.

### CODE REVIEW

File: src/Credentials/GCECredentials.php
Comment (bshaffer) [inline src/Credentials/GCECredentials.php:562]: This should only return when the client exception is a 404
```suggestion
            if ($e->hasResponse() && 404 == $e->getResponse()->getStatusCode()) {
                return self::DEFAULT_UNIVERSE_DOMAIN;
            }
            throw $e;
```

**Comment by Aungkhant884** [comment]
Patch

**Review by Aungkhant884 — APPROVED** [review]

File: src/Credentials/ServiceAccountCredentials.php
Comment (vishwarajanand) [inline src/Credentials/ServiceAccountCredentials.php:348]: Shorter?
```suggestion
        return $this->universeDomain ?: self::DEFAULT_UNIVERSE_DOMAIN;
    }
```

Comment (bshaffer) [reply src/Credentials/ServiceAccountCredentials.php:348]: I prefer the way it is for readability.

**Comment by bshaffer** [comment]
@vishwarajanand we won't cut a release until the freeze, but I'd love to get an approval on this before the EOY

File: src/GetUniverseDomainInterface.php
Comment (vishwarajanand) [inline src/GetUniverseDomainInterface.php:21]:
```suggestion
 * An interface implemented by objects that can get universe domain for Google Cloud.
```

File: src/Credentials/ServiceAccountCredentials.php
Comment (vishwarajanand) [inline src/Credentials/ServiceAccountCredentials.php:103]: It seems this is needed to preserve backwards compatibility for the case of manually created CredentialsWrapper by users.

File: src/Credentials/ServiceAccountCredentials.php
Comment (vishwarajanand) [inline src/Credentials/ServiceAccountCredentials.php:344]: Do you want to handle the case when `universeDomain` is an empty string? Or is it guaranteed to be having a valid value (eg. when always downloaded from gcp console)

**Review by vishwarajanand — APPROVED** [review]
Lgtm

File: src/Credentials/ServiceAccountCredentials.php
Comment (bshaffer) [reply src/Credentials/ServiceAccountCredentials.php:344]: Good question. The spec does not provide any guidance around empty strings for universe domain (go/cloudsdk-tpc-authlibs), except when it is a 404 error from the MDS (which is handled in phase 2).

If we added special handling for empty strings, it would be to throw an exception. But as an empty string for a universe domain will throw an error in the client libraries anyway (because we throw an error for a user-supplied empty string, and also for a mismatch), I can't think of a real important reason why we need to throw an exception now, and would prefer to leave it out (unless we can think of a specific case where it would be important to do so).

**Pull request merged**
